torch/csrc/cuda/comm.cpp

```diff
@@ -34,38 +35,75 @@ struct unique_type_checker {
-  std::vector<Tensor> tensors;
-  tensors.reserve(devices.size());
+// ***************** Broadcast *******************
+//
+// Broadcast a source tensor (CPU or CUDA) to a list of CUDA devices, or CUDA
+// tensors on one or more devices.
+
+// no checks
+static inline
+std::vector<Tensor>& _broadcast_out_impl(const Tensor& tensor, std::vector<Tensor> &out_tensors) {
 #ifdef USE_NCCL
-  if (nccl::is_available({tensor})) {
-    tensors.push_back(tensor);
-    for (auto device : devices.slice(1)) {
-      tensors.push_back(
-          at::empty(tensor.sizes(),
-          tensor.options().device(at::Device(kCUDA, device))));
-    }
-    nccl::broadcast(tensors);
+  std::vector<Tensor> nccl_list;
+  nccl_list.reserve(out_tensors.size() + 1);
+  nccl_list.push_back(tensor);
+  for (auto& out_tensor : out_tensors) {
+    nccl_list.push_back(out_tensor);
+  }
+  if (nccl::is_available(nccl_list)) {
+    nccl::broadcast(nccl_list);
   } else {
 #else
   {
 #endif
-    if (tensor.is_cuda()) {
-      tensors.push_back(tensor);
+    for (auto& out_tensor : out_tensors) {
+      out_tensor.copy_(tensor, /*non_blocking=*/true);
     }
-    IntArrayRef loop_devices = tensor.is_cuda() ? devices.slice(1) : devices;
-    for (auto device : loop_devices) {
-      tensors.push_back(tensor.to(
-          at::Device(kCUDA, device),
-          tensor.scalar_type(),
-          /*non_blocking=*/true,
-          /*copy=*/true));
+  }
+  return out_tensors;
+}
+
+std::vector<Tensor>& broadcast_out(const Tensor& tensor, std::vector<Tensor> &out_tensors) {
+  for (size_t i = 0; i < out_tensors.size(); i++) {
+    TORCH_CHECK(
+      out_tensors[i].is_cuda(),
+      "Expected all output tensors to be CUDA tensors, but output tensor at index ",
+      i, " has device '", out_tensors[i].device(), "'");
+    TORCH_CHECK(
+      out_tensors[i].sizes() == tensor.sizes(),
+      "Expected all output tensors to have same shape as the source tensor ",
+      tensor.sizes(), ", but output tensor at index ", i, " has shape ",
+      out_tensors[i].sizes());
+  }
+  return _broadcast_out_impl(tensor, out_tensors);
+}
+
+std::vector<Tensor> broadcast(const Tensor& tensor, IntArrayRef devices) {
+  std::vector<Tensor> diff_device_dst_tensors;
+  diff_device_dst_tensors.reserve(devices.size());
+  for (auto device : devices) {
+    TORCH_CHECK(
+      device >= 0,
+      "Expected non-negative device index, but got ", device);
+    if (device != tensor.get_device()) {
+      diff_device_dst_tensors.push_back(
+        at::empty(tensor.sizes(),
+          tensor.options().device(at::Device(DeviceType::CUDA, device))));  // preserve memory format
     }
   }
-  return tensors;
+  _broadcast_out_impl(tensor, diff_device_dst_tensors);
+  std::vector<Tensor> dst_tensors;
+  dst_tensors.reserve(devices.size());
+  auto it = diff_device_dst_tensors.begin();
+  for (auto device : devices) {
+    if (device != tensor.get_device()) {
+      dst_tensors.push_back(*it++);
+    } else {
```

Comment:
I might miss sth, but it doesn't seem this `else` branch will ever be reached? This function does not add the input `tensor` to `diff_device_dst_tensors`, and it seems neither does `_broadcast_out_impl`?

Reply:
If the target device is the same as the source device, we don't broadcast for that device (see line 88 above) and just return the source tensor (var `tensor`) here since there was no need to move.

Reply:
ah I see